edx/edx-enterprise
# feat: Use updated template model for bulk enrollment ENT-4322

This approach is basically to add a`template_type` column to the model holding email templates.
Also removes use of file based templates to single source templates from db.

It allows us to single source templates in one place - the database

Also a data migration added to populate default template for enterprise_customer=None, for both template_type values (SELF_ENROLL and ADMIN_ENROLL). The latter is used by bulk enrollment emails. The former template is used by by any other calls to the `notify_learners()` function that does NOT specify `admin_enrollment=True`.


__Basic approach:__

We now fetch the model with type='ADMIN_ENROLL' if bulk enroll case is specified, otherwise we default to the 'SELF_ENROLL' (or default) template (templates fetched from the database table for `EnrollEmailNotificationTemplate` model).

In case template override per customer is not found in the db for the given template_type, we use the fallback template from the db where enterprise_customer=None and type=Desired_Type 

The `admin_enrollment=True` switch passed to `notify_learners` (with default value of False) instructs the code to pick up the admin enrollment template if it exists in db, or from the admin enrollment template db records, as a fallback.

Also we log warning when we cannot find a template:

e.g.,
```
WARNING  enterprise.utils:utils.py:497 Cannot find email templates for EdX, template_type: DEFAULT_ENROLL. Not sending notification email.
```

Tests added to cover template finding logic

### Manual testing notes

_with no templates at all_: I only get a logging message as above, and no email attempts made according to email log file

_with only learner initiated default enroll template but no admin default template, still no per enterprise templates_: again, for bulk enroll case, no attempt to email was made, only logging seen. But if I remove the `admin_enrollment=true` to switch modes, I now see it picking up the self-enroll template instead

_with default templates for both learner and admin initiated templates_: for bulk enroll, the correct template was chosen

in all cases, checked the /edx/src/ace_messages in the lms container.

__Email looks like this__

<img width="922" alt="Screen Shot 2021-06-08 at 10 19 18 PM" src="https://user-images.githubusercontent.com/528166/121282534-acda1880-c8a7-11eb-8266-6582318eb374.png">

My local django admin vie with three templates (two from migration default):

<img width="872" alt="Screen Shot 2021-06-09 at 8 27 39 AM" src="https://user-images.githubusercontent.com/528166/121354167-a033df80-c8fc-11eb-9009-8447cb6a473e.png">


ENT-4322

**Merge checklist:**
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.

**Post merge:**
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.
